R/utilities-patterns.R

```diff
@@ -0,0 +1,102 @@
+#'
+#' @return A `character` vector of colours, or list of `<GridPattern>` objects.
+#' @export
+#'
+#' @examples
+#' # Typical colour input
+#' fill_alpha("red", 0.5)
+#'
+#' # Pattern input
+#' fill_alpha(list(linearGradient()), 0.5)
+fill_alpha <- function(fill, alpha) {
+  if (!is.list(fill)) {
+    # Happy path for no patterns
+    return(alpha(fill, alpha))
+  }
+  if (any(vapply(fill, is_pattern, logical(1)))) {
+    if (utils::packageVersion("grid") < "4.2") {
+      # Pattern fills were introduced in R 4.1.0, but *vectorised* patterns
+      # were only introduced in R 4.2.0.
+      cli::cli_abort(
+        "Using patterns in {.pkg ggplot2} requires at least R version 4.2.0."
+      )
+    }
+    fill <- pattern_alpha(fill, alpha)
+    return(fill)
+  } else {
+    # We are either dealing with faulty fill specification, or we have a legend
+    # key that is trying to draw a single colour. It can be given that colour
+    # as a list due to patterns in other keys.
+    msg <- paste0(
+      "{.field fill} must be a valid colour or list of ",
+      "{.cls GridPattern} objects."
+    )
+    # If single colour list, try applying `alpha()`
+    fill <- try_fetch(
+      Map(alpha, colour = fill, alpha = alpha),
+      error = function(cnd) {
+        cli::cli_abort(msg, call = expr(fill_alpha()))
+      }
+    )
+    # `length(input)` must be same as `length(output)`
+    if (!all(lengths(fill) == 1)) {
+      cli::cli_abort(msg)
+    }
+    return(unlist(fill))
+  }
+}
+
+# Similar to grid:::is.pattern
+is_pattern <- function(x) {
+  inherits(x, "GridPattern")
+}
+
+# Function that applies alpha to <GridPattern> objects.
+# For linear or radial gradients, this is as simple as modifying their `colours`
+# slot with an alpha.
+# For tiled patterns, we attach an alpha mask in the grobs' viewport.
+pattern_alpha <- function(x, alpha) {
+  if (!is.list(x)) {
+    # If this is a plain colour, convert to pattern because grid doesn't accept
+    # mixed patterns and plain colours.
+    out <- pattern(rectGrob(), gp = gpar(fill = alpha(x, alpha)))
+    return(out)
+  }
+  if (!is_pattern(x)) {
+    out <- Map(pattern_alpha, x = x, alpha = alpha)
+    return(out)
+  }
+  if (inherits(x, c("GridLinearGradient", "GridRadialGradient"))) {
+    # Apply alpha to gradient colours
+    x$colours <- alpha(x$colours, alpha[1])
+    return(x)
+  }
+  no_alpha <- is.na(alpha[1]) || alpha[1] == 1
+  if (inherits(x, "GridTilingPattern") && !no_alpha) {
+    # Dig out the grob from the function environment
+    grob <- env_get(environment(x$f), "grob")
+    # Apply a mask in the grob's viewport
+    mask <- as.mask(rectGrob(gp = gpar(fill = alpha("black", alpha[1]))))
+    if (is.null(grob$vp)) {
+      grob$vp <- viewport(mask = mask)
+    } else {
+      grob$vp$mask <- mask
+    }
+    # Re-attach new function environment
+    new_env <- new.env(parent = environment(x$f))
+    env_bind(new_env, grob = grob)
+    environment(x$f) <- new_env
```

Comment:
This 'feels' dirty, but I can't see how to apply an alpha to tiled patterns otherwise.

Reply:
An alternative may be instead of extracting the `grob` you extract the `gp` and update the alpha in that `gpar()` object...  If you do extract the `grob` you could also do `grob = grobTree(grop, gp = gpar(alpha = ...`

If you are going to alpha mask it is a good idea to make sure the color choice works if the alpha mask is mis-interpreted as a luminance mask (i.e. the popular program `inkscape` does such a mis-interpretation).  Maybe "black" works but maybe "white" is a better choice.

Reply:
I think the 'dirty' part is rummaging through function environments to make edits. Setting `gpar(alpha = ...)` works for some grobs, but unfortunately not for others, like `rasterGrob()`. Using "white" is a good suggestion, I hadn't quite considered this choice.

Reply:
If using alpha masking for "clipping" white is definitely better than black.  Not sure for non-1 alpha values.  Perhaps instead of "white" it could be that `rgb(alpha, alpha, alpha)` is better?  https://github.com/r-lib/svglite/issues/146

Reply:
Just tried this, and it sort of makes sense for inkscape, but it doesn't quite correspond to what one would see with e.g. png exports. I'm a bit worried that with the `rgb(alpha, alpha, alpha)` approach we'd be fooling users into thinking their plot has rendered correctly in inkscape (which, as you've pointed out, it doesn't because it uses luminance masks).

Reply:
* Here is some documentation saying that the alpha value for a luminance mask should be equal to the luminance value times the alpha value so `alpha("white", alpha[1])` should theoretically work (i.e. alpha = 1 * alpha): https://css-tricks.com/almanac/properties/m/mask-mode/#aa-luminance-masks
* Playing around this seems to work with my version of `inkscape` (and it seems that `eog` was improved to properly interprets alpha masks as alpha masks).

Reply:
For rasterGrob patterns, we could perhaps just multiply the transparency of the colours directly. The `rgb(alpha, alpha, alpha)` is probably the most sensible choice, as the luminance of that grayscale is not exactly but roughly equal to the input values.

Reply:
> For rasterGrob patterns, we could perhaps just multiply the transparency of the colours directly.

> For rasterGrob patterns, we could perhaps just multiply the transparency of the colours directly.

This is what `{gridpattern}` / `{ggpattern}` basically does for "array" style patterns.

> The rgb(alpha, alpha, alpha) is probably the most sensible choice, as the luminance of that grayscale is not exactly but roughly equal to the input values.

Are you sure `rgb(alpha, alpha, alpha)` works better than `alpha("white", alpha[1])` if the luminance mask value transparency is the luminosity value times the alpha value?  `alpha("white", alpha[1])` should have an equivalent alpha mask value and luminance mask value (alpha = 1 * alpha) whereas `rgb(alpha, alpha, alpha)` does not?

Reply:
@thomasp85 ggplot2 doesn't import {farver} but for alpha multiplication, we'd need something like `farver::multiply_channel(x, "alpha", 0.5)`. Would it make sense to wrap this functionality in the {scales} package, that already imports farver?